### PR TITLE
updated src and UT files for implementation of "selectively ignore some measure"

### PR DESCRIPTION
**Pull request by Madhu-NEC**
Issue #416  implemented.

### CODE REVIEW

**Comment by Madhu-NEC** [comment]
Two test cases of HTTP Measure are failing due to below reason:

In the execution of test cases, the request for device provisioning at line [211](https://github.com/Madhu-NEC/iotagent-json/blob/selectively_ignore_some_measure/test/unit/ngsiv2/HTTP_reveice_measures-test.js#L211)  has been done which uses node-lib repo. But the source code of node-lib has also been modified for this feature. When the travis is running this PR, the updated code of node-lib is not reflected.

Note: All test cases are running in my local environment as I have updated code of both the repos in my local environment.

**Comment by fgalan** [comment]
> Two test cases of HTTP Measure are failing due to below reason:
> 
> In the execution of test cases, the request for device provisioning at line [211](https://github.com/Madhu-NEC/iotagent-json/blob/selectively_ignore_some_measure/test/unit/ngsiv2/HTTP_reveice_measures-test.js#L211) has been done which uses node-lib repo. But the source code of node-lib has also been modified for this feature. When the travis is running this PR, the updated code of node-lib is not reflected.
> 
> Note: All test cases are running in my local environment as I have updated code of both the repos in my local environment.

You can try to adjust the packages.json dependency in this PR. Something like this I guess:

```
     "iotagent-node-lib": "git://github.com/Madhu-NEC/iotagent-node-lib.git#selectively_ignore_some_measure", 
```

That would make the code in this PR to use your modifications also in the PR https://github.com/telefonicaid/iotagent-node-lib/pull/835. Once both PRs gets at LGTM point, the procedure will be as follows:

1. https://github.com/telefonicaid/iotagent-node-lib/pull/835 will be merged
2. packages.json in this PR will be modified but to "iotagetnet-node-lib" to "git://github.com/telefonicaid/iotagent-node-lib.git#master"
3. Check travis is ok
3. PR will be merged

**Comment by Madhu-NEC** [comment]
> You can try to adjust the packages.json dependency in this PR. Something like this I guess:

Hi @fgalan 
I have updated the package.json. Now all the test cases are passing.

**Comment by Madhu-NEC** [comment]
Hi @fgalan ,
Please provide your feedback. 
If it is ok then please merge the PR.

**Comment by fgalan** [comment]
> 
> 
> Hi @fgalan ,
> Please provide your feedback.
> If it is ok then please merge the PR.

I'm currently out of work due to Christmas Holidays. I'll have a look when I'll back. Sorry for the inconveniences.

File: CHANGES_NEXT_RELEASE [hunk 1]
```diff
@@ -1,2 +1,3 @@
+- Added feature: selectively ignore some of the measures (#416)
```
Comment (fgalan) [inline CHANGES_NEXT_RELEASE]:
```suggestion
Add: explicitAttrs flag (in configuration and also group/device provisioning) to progress only the measures corresponding to attributes declared in attributes array (#416)
```

to align with a similar text in the associated PR in the ioagent-node-lib

File: config.js [hunk 2]
```diff
@@ -258,7 +258,11 @@ config.iota = {
      * Default resource of the IoT Agent. This value must be different for every IoT Agent connecting to the IoT
      * Manager.
      */
-    defaultResource: '/iot/json'
+    defaultResource: '/iot/json',
+    /**
+     * flag indicating whether the incoming measures to the IoTAgent should be processed as per the "attributes" field.
+     */
+    defaultExplicitAttrs: true
```
Comment (fgalan) [inline config.js]:
```suggestion
    explicitAttrs: true
```

as suggested in https://github.com/telefonicaid/iotagent-node-lib/pull/835/files#r363794493

File: package.json [hunk 3]
```diff
@@ -61,7 +61,7 @@
     "body-parser": "1.18.3",
     "dateformat": "3.0.3",
     "express": "~4.16.4",
-    "iotagent-node-lib": "git://github.com/telefonicaid/iotagent-node-lib.git#master",
+    "iotagent-node-lib": "git://github.com/Madhu-NEC/iotagent-node-lib.git#selectively_ignore_some_measure",
```
Comment (fgalan) [inline package.json]: To be revered once PR https://github.com/telefonicaid/iotagent-node-lib/pull/835 gets merged (just a remark note).

File: test/unit/ngsiv2/HTTP_reveice_measures-test.js [hunk 4]
```diff
@@ -182,6 +183,79 @@ describe('HTTP: Measure reception ', function() {
...
+
+        it('should send its value to the Context Broker', function(done) {
+            request(optionsMeasure, function(error, result, body) {
+                contextBrokerMock.done();
+                done();
+            });
+        });
+    });
+
+    describe('When a POST measure arrives with a TimeInstant attribute in the body for a device with explicitAttrs:true', function() {
```
Comment (fgalan) [inline test/unit/ngsiv2/HTTP_reveice_measures-test.js]: For the sake of simetry, the parallel describe in line 158 should be:

```
     describe('When a POST measure arrives with a TimeInstant attribute in the body for a device with explicitAttrs:false', function() {
```

isn't it?

File: test/unit/ngsiv2/HTTP_reveice_measures-test.js [hunk 5]
```diff
@@ -254,6 +329,79 @@ describe('HTTP: Measure reception ', function() {
...
+
+        it('should send its value to the Context Broker', function(done) {
+            request(optionsMeasure, function(error, result, body) {
+                contextBrokerMock.done();
+                done();
+            });
+        });
+    });
+
+    describe('When a POST measure arrives with a TimeInstant query parameter in the body for a device with explicitAttrs:true', function() {
```
Comment (fgalan) [inline test/unit/ngsiv2/HTTP_reveice_measures-test.js]: Similar to a previous comment, for simmetry, the describe in line 304 should be changed to:

```
describe('When a POST measure arrives with a TimeInstant query parameter in the body with explicitAttrs:false', function() {
```

**Comment by fgalan** [comment]
Except for some minor comments (that I hope can be easily solved), the PR is almost ready. Thus, I wonder if it could be possible to do a "parallel" PR for IOTA-UL so we can merge the three PR (IOTA Lib, IOTA-JSON and IOTA-UL) at the same time.

That would ensure that IOTA-UL and IOTA-JSON agents evolved homogenously, which is pretty important from a usage point of view.

What do you think?

File: CHANGES_NEXT_RELEASE (hunk 1 above)
Comment (Madhu-NEC) [reply CHANGES_NEXT_RELEASE]: Fixed in 83834fe

File: config.js (hunk 2 above)
Comment (Madhu-NEC) [reply config.js]: Fixed in 4724c32

File: test/unit/ngsiv2/HTTP_reveice_measures-test.js (hunk 4 above)
Comment (Madhu-NEC) [reply test/unit/ngsiv2/HTTP_reveice_measures-test.js]: Fixed in 4724c32

File: test/unit/ngsiv2/HTTP_reveice_measures-test.js (hunk 5 above)
Comment (Madhu-NEC) [reply test/unit/ngsiv2/HTTP_reveice_measures-test.js]: Fixed in 4724c32

**Comment by Madhu-NEC** [comment]
Hi @fgalan ,
I have updated the PR for all comments and request you to please verify the same.
If it is ok please merge the PR.

File: config.js [hunk 6]
```diff
@@ -258,7 +258,11 @@ config.iota = {
      * Default resource of the IoT Agent. This value must be different for every IoT Agent connecting to the IoT
      * Manager.
      */
-    defaultResource: '/iot/json'
+    defaultResource: '/iot/json',
+    /**
+     * flag indicating whether the incoming measures to the IoTAgent should be processed as per the "attributes" field.
+     */
+    explicitAttrs: true
```
Comment (fgalan) [inline config.js]:
```suggestion
    explicitAttrs: false
```

In order to keep backward compatibility in previous versions of the IOTA.

**Comment by fgalan** [comment]
In which test can I see the effect of the implemented functionality? I mean, something like this:

1. Device provisioned with attributes A and B
2. Measure send: B and C
3. Result: only B progresses to CB

**Comment by fgalan** [comment]
> Hi @fgalan ,
> I have updated the PR for all comments and request you to please verify the same.
> If it is ok please merge the PR.

Thanks for the PR update! I have provided some additional feedback (one minor comment and a question/doubt).

What about the implementation of the same feature for IOTA-UL? It would be great if you could do the PR also there (I understand it is basically copying the same changes in the other agents) so we can merge both together. It is not desirable to have the agents unbalanced... (one with a feature that the other doesn't have).

File: config.js (hunk 6 above)
Comment (Madhu-NEC) [reply config.js]: Fixed in f8d801e

**Comment by Madhu-NEC** [comment]
> In which test can I see the effect of the implemented functionality? I mean, something like this:
> 
> 1. Device provisioned with attributes A and B
> 2. Measure send: B and C
> 3. Result: only B progresses to CB

Hi @fgalan ,
Please refer the below test case which is written in file **test/unit/ngsiv2/HTTP_reveice_measures-test.js**:

**"describe('When a POST measure arrives with a TimeInstant attribute in the body for a device with explicitAttrs:true", function() {"**

**Comment by Madhu-NEC** [comment]
> What about the implementation of the same feature for IOTA-UL? It would be great if you could do the PR also there (I understand it is basically copying the same changes in the other agents) so we can merge both together. It is not desirable to have the agents unbalanced... (one with a feature that the other doesn't have).

Yes, We will implement the same feature for IOTA-UL also. Meanwhile I request to merge this [PR ](https://github.com/telefonicaid/iotagent-json/pull/458/)but we can keep the issue #416  open until the PR for IOTA-UL is raised and merged.

**Comment by fgalan** [comment]
> Yes, We will implement the same feature for IOTA-UL also. Meanwhile I request to merge this [PR ](https://github.com/telefonicaid/iotagent-json/pull/458/)but we can keep the issue #416 open until the PR for IOTA-UL is raised and merged.

I think is preferable to leave the PRs on iotagent-nodel-lib and IOTA-JSON opened while you provide the PR on IOTA-UL. While you prepare the contribution for IOTA-UL some issue could arise, involving fixes in the previous PRs. In addition, it will be easier to regression both agents (IOTA-UL and IOTA-JSON) in our e2e CI system before merging any one of them.

Do you agree?

**Comment by Chandradeep-NEC** [comment]
@fgalan 
I agree with your concern and suggestion and also i do appreciate your review and responses.
We have one concern as currently we are busy in other work, so may be it'll take time to provide patch for feature #416  in IOTA-UL. 
I worry as we have already put huge efforts for supporting of this feature #416  in IOTA-JSON and IOTA-Node-lib but this PR will still be open due to IOTA-UL.
I request you to merge PR for IOTA-JSON and IOTA-Node-lib. We'll provide some notes in document of IOTA-UL regarding feature #416  and once we'll create PR then will remove the note from documentation.

**Comment by fgalan** [comment]
Don't worry, your valuable efforts are not going to be lost :)

The PRs on IOTA-Node-Lib and IOTA-JSON can still while you get time to provide de IOTA-UL PR. That PR should be pretty straightforward, as it is basically "copying" the one from IOTA-JSON (which is small: only +236 -15 lines of changes). Maybe some other could help here...

**Comment by fgalan** [comment]
With the creation of PR https://github.com/telefonicaid/iotagent-ul/pull/430 (which I understand completes the selectively ignore some measures functionality along all the main agents) I think is the moment of updating this PR with master and solve the current conflict.

@Madhu-NEC could you manage? Thanks!

**Comment by Madhu-NEC** [comment]
>@Madhu-NEC could you manage? Thanks!

Hi @fgalan , I have updated the PR with master. Also, **node-lib dependency** is adjusted to [PR#873](https://github.com/telefonicaid/iotagent-node-lib/pull/873).
Thanks.

**Comment by fgalan** [comment]
Now that https://github.com/telefonicaid/iotagent-node-lib/pull/873 has been merged, packages.json in this PR can be changed to master.

**Comment by Madhu-NEC** [comment]
> Now that [telefonicaid/iotagent-node-lib#873](https://github.com/telefonicaid/iotagent-node-lib/pull/873) has been merged, packages.json in this PR can be changed to master.

Hi @fgalan ,
I have updated the package.json to master.
Thanks!

**Pull request merged**